flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/OperatorCoordinatorHolder.java

```diff
@@ -301,6 +308,43 @@ private void checkpointCoordinatorInternal(
         }
     }
 
+    private void completeCheckpointOnceEventsAreDone(
+            final long checkpointId,
+            final CompletableFuture<byte[]> checkpointFuture,
+            final byte[] checkpointResult) {
+
+        final Collection<CompletableFuture<?>> pendingEvents =
+                unconfirmedEvents.getCurrentIncompleteAndReset();
+        if (pendingEvents.isEmpty()) {
+            checkpointFuture.complete(checkpointResult);
+            return;
+        }
+
+        LOG.info(
+                "Coordinator checkpoint {} for coordinator {} is awaiting {} pending events",
+                checkpointId,
+                operatorId,
+                pendingEvents.size());
+
+        final CompletableFuture<?> conjunct = FutureUtils.waitForAll(pendingEvents);
```

Comment:
For self backlog(not reviewing): This should still work only for source tasks. This guarantees that checkpoint barrier for all tasks will come after events before checkpoint barrier injection. But for no source tasks, new events could arrive before checkpoint barrier from upstream.

Reply:
Still self backlog, for solely source exactly-once, this do more than necessary.